### ISSUE TITLE
RFC: Command Aliases

### ISSUE THREAD

**Issue by lukeed** [lukeed/sade#26]
Command aliases are (generally) an abbreviated form of the command. They're typically geared for lazy typers and/or power users 😉 

The most common example that many of us are probably familiar with is `npm install`, which is also accessible via `npm i`. All options and arguments still carry over identically. 

Here's how it would/could be added in a Sade program:

```js
sade('npm')
  .command('install [package]', 'Install a package', { alias: 'i' })
  .option('-P, --save-prod', 'Package will appear in your dependencies.')
  .option('-D, --save-dev', 'Package will appear in your devDependencies.')
  .option('-O, --save-optional', 'Package will appear in your optionalDependencies')
  .option('-E, --save-exact', 'Save exact versions instead of using a semver range operator')
  .action(handler);
```

When running `npm --help`, you'll see this:

```
  Usage
    $ npm <command> [options]

  Available Commands
    install    Install a package

  For more info, run any command with the `--help` flag
    $ npm install --help

  Options
    -v, --version    Displays current version
    -h, --help       Displays this message
```
> **Note:** There's no mention of aliases here. This output is unchanged

And when running `npm install --help` you'd see this:

```
  Description
    Install a package

  Usage
    $ npm install [package] [options]

  Aliases
    $ npm i

  Options
    -P, --save-prod        Package will appear in your dependencies.
    -D, --save-dev         Package will appear in your devDependencies.
    -O, --save-optional    Package will appear in your optionalDependencies
    -E, --save-exact       Save exact versions instead of using a semver range operator
    -h, --help             Displays this message
```

Of course, running `npm install` and `npm i` are synonymous, which also means that running `npm i -h` would also print the above output.

---

And now, some options and voting 😄 

### 1. Do we even want this?

[![](https://api.gh-polls.com/poll/01DE373GJT1NQ218A6J3B7JPQE/Yes!)](https://api.gh-polls.com/poll/01DE373GJT1NQ218A6J3B7JPQE/Yes!/vote)
[![](https://api.gh-polls.com/poll/01DE373GJT1NQ218A6J3B7JPQE/No%2C%20do%20not%20add%20this.)](https://api.gh-polls.com/poll/01DE373GJT1NQ218A6J3B7JPQE/No%2C%20do%20not%20add%20this./vote)
[![](https://api.gh-polls.com/poll/01DE373GJT1NQ218A6J3B7JPQE/No%2C%20but%20I%20don't%20really%20care.)](https://api.gh-polls.com/poll/01DE373GJT1NQ218A6J3B7JPQE/No%2C%20but%20I%20don't%20really%20care./vote)

### 2. Should multiple aliases be allowed?

[![](https://api.gh-polls.com/poll/01DE36GDAAGWYR84SKD3H6ATPK/Allow%20multiple%20aliases%20per%20command)](https://api.gh-polls.com/poll/01DE36GDAAGWYR84SKD3H6ATPK/Allow%20multiple%20aliases%20per%20command/vote)
[![](https://api.gh-polls.com/poll/01DE36GDAAGWYR84SKD3H6ATPK/Allow%20only%20one%20alias%20per%20command)](https://api.gh-polls.com/poll/01DE36GDAAGWYR84SKD3H6ATPK/Allow%20only%20one%20alias%20per%20command/vote)

Back to the `npm install` example – they actually have 3 aliases: `i, isntall, add`<br>
For the same Sade program, this would mean the `alias` option accepts an array:

```js
prog.command('install [package]', '...', { alias: ['i', 'add', 'isntall'] })
```

```
  Description
    Install a package

  Usage
    $ npm install [package] [options]

  Aliases
    $ npm i
    $ npm add
    $ npm isntall

  Options
    -P, --save-prod        Package will appear in your dependencies.
    -D, --save-dev         Package will appear in your devDependencies.
    -O, --save-optional    Package will appear in your optionalDependencies
    -E, --save-exact       Save exact versions instead of using a semver range operator
    -h, --help             Displays this message
```

### 3. Should aliases be added through a method?

[![](https://api.gh-polls.com/poll/01DE36TCR4K7F8YZ6C1VHXBVRJ/Add%20through%20%60command()%60%20option)](https://api.gh-polls.com/poll/01DE36TCR4K7F8YZ6C1VHXBVRJ/Add%20through%20%60command()%60%20option/vote)
[![](https://api.gh-polls.com/poll/01DE36TCR4K7F8YZ6C1VHXBVRJ/Add%20through%20new%20%60alias()%60%20method)](https://api.gh-polls.com/poll/01DE36TCR4K7F8YZ6C1VHXBVRJ/Add%20through%20new%20%60alias()%60%20method/vote)

Personally, a new `.alias()` method would _only_ make sense if we allow multiple aliases. Otherwise it'd be awkward and a no-go from me, I think.

Another drawback is that this would be the only API method that _requires_ declaration _after_ a `.command()` usage. Unlike `.option()`, you can't declare a "global alias" since that means nothing. 

A final drawback is that a command's aliases is declared across two methods, whereas an option's alias is declared in one fell swoop. Keeping command aliases within the `command()` options will keep it "in one fell swoop"

```js
// Proposed:
prog.command('install [package]', '...', { 
  alias: ['i', 'add', 'isntall']
})

// Alternative:
prog
  .command('install [package]', '...')
  .alias('i', 'add', 'isntall')
```  

---

Thanks for dropping by!

PS: Votes are anonymous 👍

**Comment by tunnckoCore** [comment]
Seems like people want it, haha. Interesting.

> Personally, a new .alias() method would only make sense if we allow multiple aliases.

Agree. I'm for option to the `.command()` method.

```js
// yarn
// yarn install
// yarn add
// yarn i
sade('yarn')
  .command('install [package]', '...', {
    alias: ['i', 'add', 'isntall'],
    default: true
  })
```

**Comment by tunnckoCore** [comment]
I've done it. Want a PR?

**Comment by tunnckoCore** [comment]
Will implement and `.alias` method too.

**Comment by lukeed** [comment]
No thank you, but I appreciate it! I should have mentioned in the original message that I already have all this work completed.

I am still waiting for feedback. I've been checking back on this every so often.

It looks like the NOs are still winning, and even then, there aren't clear winners on the remaining questions.

Will try to get more votes on this next week, starting tomorrow.

IMO, not good to add stuff that majority doesn't want. Feature creep leads to bloat/clutter 😬

**Comment by fraxken** [comment]
hi, 

Is there is way to implement alias for a command right now ? The subject interests me a lot because me and my team we run almost ~10 CLI with Sade and we are waiting a lot after this RFC.

Example here:
https://github.com/SlimIO/CLI/blob/master/bin/index.js

Like npm we have command like **add** , **init** and **remove** :\ 

#team-alias 😂 

Best Regards,
Thomas

**Comment by tunnckoCore** [comment]
Want to believe that you can, currently, with Sade. But still thinking it how. :laughing: 

I'm thinking from few days to publish `@hela/sade` or `@tunnckocore/sade` not "just because" but as an early level experimentation layer and because I really building things on top of these things and can wait - some things get implemented/merged some not, and that's totally normal :)

Kind of like Fedora to RedHat :D

**Comment by lukeed** [comment]
I'm leaning towards shipping this. Still waiting for feedback in case there's strong opposition.

Sorry I forgot to shop for feedback last week, but I [just did it now](https://twitter.com/lukeed05/status/1193968946392657920).

There _is_ a way to force it in there right now through method overrides, but it's pretty messy and not something I want to recommend haha. And it'd list aliases as if they were separate commands in the help text, which is confusing.

We'll (very likely) have the much nicer, polished solution in the next minor release.

**Comment by tunnckoCore** [comment]
It's actually easy. The thing is you won't get correct/good help output, since you need changes in the `.parse` method too.

~~It's along these lines~~ Gave up.

**Comment by tunnckoCore** [comment]
Update?

**Comment by lukeed** [comment]
Hey sorry - honestly I forgot about this while swamped with work.

What I have now allows both the method and the command option. The `alias()` will throwing air if you try to use it before having a command context defined.

It will also throw an error if you try to define an alias while within "single" mode, because that makes no sense // isn't applicable.

I just have to add a bunch of tests now. I added it to some local programs and it fuels really nice.

Now I just have to decide if Sade should explicitly check and prevent overlaps between aliases and other commands & vice versa. For example, I can't add a `i` alias for `install` if there's already another command/alias for `i` -- similarly, I can't add an `i` command if it's already aliased elsewhere.

Should I add that or trust the dev knows what they're doing? 

it sort of feels like I _should_ add it, but at the same time I'm a little bit lazy and inclined to say that you should keep track 😬

**Comment by tunnckoCore** [comment]
> because that makes no sense // isn't applicable.

Yea, totally I just added that before few hours too :D I actually reimplemented the whole Sade from scratch, just for the sake of seeing where it can go, and because mroe and more needs.

> I just have to add a bunch of tests now.

Yea... I didn't wrote any tests last year for anything :rofl: Same way... I use all of my things so... i immediately see what is happening. Kind of miss those TDD times :D 

Okay, why not push a branch and add dist-tag and we can continue.

> Should I add that or trust the dev knows what they're doing?

Trust. Otherwise it will add even more amount of code in codebase that is already hard to follow (for me, just because the code style) and already had some problems with it, and that's why i rewrote it.

**Comment by lukeed** [comment]
PR open, will publish tomorrow unless something major comes up in feedback.

I went with the "developer beware" approach, which does always leave room for a non-breaking patch release if people want the checks in place.

> I actually reimplemented the whole Sade from scratch

Maybe you'll drop your fork now, haha. Otherwise I'm confused as to why you're leading the charge on this feature. Also, too late now, but I would have appreciated an _actual_ fork if you planned on doing so. I see & appreciate the NOTICES/other mentions, but there's still [a lack of continuity and history there](https://github.com/tunnckoCore/hela/tree/ab6eb4f1dda226781defd99967c4dc971018f09f/packages/sade/src). Not a big deal, just a sidenote 👍

**Comment by tunnckoCore** [comment]
Yea, most probably I'll drop it. It's one of the "major" features that I started it for, besides few other things.

The reimplementation grow from the "patch" that I maintained some time at https://github.com/tunnckoCore/opensource/tree/master/patches/sade. I, sure, could "actual fork" it, but just was more comfortable for me to be at the `opensource` monorepo. In same time there is also copy in the `hela` ;d Then after more usage few other things appeared, and I have problem specifically with the codebase/codestyle. Last night i said fack that monorepo approach and all the things (i'm seriously sick of all the back and forth of monorepos, automation, building and etc, i'm really doing nothing whole last year because of this, dramatic decrease of productivity ;d), and joined all the separate packages (same as a monorepo without being a monorepo, haha - brilliant, heck why i need monorepo for this..) and finally published hela@next (v3) ;d which includes the rewrite of Sade as folder called `yaro` as a temporary thing :D But it seems i didn't pushed it. If you are curious and have time you can install it and check, it's the same size without the help handling :)

Aaaaaaaaanyway... long story short, I'll return to Sade soon.

**Comment by lukeed** [comment]
No worries! Sorry didn't mean for you to try to defend yourself, my bad. I actually dunno how you'd do it within a monorepo
